File: salesloft_client.rb
require 'net/http'
require 'json'
require 'uri'

class SalesloftClient
  BASE_URI = 'https://api.salesloft.com'
  API_PATH = '/v2'

  def initialize(api_key)
    @api_key = api_key
  end

  def me
    get('me')
  end

  def get_person(id)
    get("people/#{id}")
  end

  def create_person(data)
    post('people', data)
  end

  def update_person(id, data)
    put("people/#{id}", data)
  end

  def list_people(params = {})
    get('people', params)
  end

  def cadences
    get('cadences')
  end

  def add_person_to_cadence(person_id, cadence_id)
    post("cadence_memberships", { person_id: person_id, cadence_id: cadence_id })
  end

  def person_url(id)
    "#{BASE_URI}/app/people/#{id}"
  end

  def cadence_people_url(cadence_id)
    "#{BASE_URI}/app/cadences_v2/#{cadence_id}/v2/cadence_people"
  end

  private

  def headers
    {
      'Authorization' => "Bearer #{@api_key}",
      'Content-Type' => 'application/json',
      'Accept' => 'application/json'
    }
  end

  def get(path, params = {})
    uri = URI("#{BASE_URI}#{API_PATH}/#{path}.json")
    uri.query = URI.encode_www_form(params) unless params.empty?
    request = Net::HTTP::Get.new(uri)
    headers.each { |k, v| request[k] = v }
    execute(uri, request)
  end

  def post(path, data = {})
    uri = URI("#{BASE_URI}#{API_PATH}/#{path}.json")
    request = Net::HTTP::Post.new(uri)
    headers.each { |k, v| request[k] = v }
    request.body = data.to_json
    execute(uri, request)
  end

  def put(path, data = {})
    uri = URI("#{BASE_URI}#{API_PATH}/#{path}.json")
    request = Net::HTTP::Put.new(uri)
    headers.each { |k, v| request[k] = v }
    request.body = data.to_json
    execute(uri, request)
  end

  def execute(uri, request)
    http = Net::HTTP.new(uri.host, uri.port)
    http.use_ssl = true
    http.open_timeout = 15
    http.read_timeout = 30

    response = http.request(request)

    case response
    when Net::HTTPSuccess
      JSON.parse(response.body)
    else
      { 'error' => response.code, 'message' => response.message, 'body' => response.body }
    end
  end
end

File: salesloft-mcp-server/src/salesloft-client.ts
import https from "node:https";

const BASE_URL = "https://api.salesloft.com";

export interface SalesloftConfig {
  apiKey: string;
}

export interface ApiResponse {
  data?: any;
  metadata?: {
    paging?: {
      per_page: number;
      current_page: number;
      next_page: number | null;
      prev_page: number | null;
      total_pages: number;
      total_count: number;
    };
  };
  errors?: any[];
}

export class SalesloftClient {
  private apiKey: string;

  constructor(config: SalesloftConfig) {
    this.apiKey = config.apiKey;
  }

  private request(
    method: string,
    path: string,
    body?: Record<string, any>,
    queryParams?: Record<string, string>
  ): Promise<ApiResponse> {
    return new Promise((resolve, reject) => {
      let url = `${BASE_URL}/v2/${path}.json`;
      if (queryParams) {
        const params = new URLSearchParams(queryParams);
        const qs = params.toString();
        if (qs) url += `?${qs}`;
      }

      const parsed = new URL(url);
      const options: https.RequestOptions = {
        hostname: parsed.hostname,
        port: 443,
        path: parsed.pathname + parsed.search,
        method,
        headers: {
          Authorization: `Bearer ${this.apiKey}`,
          "Content-Type": "application/json",
          Accept: "application/json",
        },
      };

      const req = https.request(options, (res) => {
        let data = "";
        res.on("data", (chunk) => (data += chunk));
        res.on("end", () => {
          try {
            if (res.statusCode && res.statusCode >= 200 && res.statusCode < 300) {
              resolve(data ? JSON.parse(data) : {});
            } else {
              const parsed = data ? JSON.parse(data) : {};
              reject(
                new Error(
                  `Salesloft API error ${res.statusCode}: ${JSON.stringify(parsed)}`
                )
              );
            }
          } catch {
            reject(new Error(`Failed to parse response: ${data}`));
          }
        });
      });

      req.on("error", reject);
      req.setTimeout(30000, () => {
        req.destroy();
        reject(new Error("Request timed out"));
      });

      if (body && (method === "POST" || method === "PUT" || method === "PATCH")) {
        req.write(JSON.stringify(body));
      }
      req.end();
    });
  }

  // ── Current User ──
  async getMe(): Promise<ApiResponse> {
    return this.request("GET", "me");
  }

  // ── People ──
  async listPeople(params?: Record<string, string>): Promise<ApiResponse> {
    return this.request("GET", "people", undefined, params);
  }

  async getPerson(id: number): Promise<ApiResponse> {
    return this.request("GET", `people/${id}`);
  }

  async createPerson(data: Record<string, any>): Promise<ApiResponse> {
    return this.request("POST", "people", data);
  }

  async updatePerson(
    id: number,
    data: Record<string, any>
  ): Promise<ApiResponse> {
    return this.request("PUT", `people/${id}`, data);
  }

  async deletePerson(id: number): Promise<ApiResponse> {
    return this.request("DELETE", `people/${id}`);
  }

  // ── Accounts ──
  async listAccounts(params?: Record<string, string>): Promise<ApiResponse> {
    return this.request("GET", "accounts", undefined, params);
  }

  async getAccount(id: number): Promise<ApiResponse> {
    return this.request("GET", `accounts/${id}`);
  }

  async createAccount(data: Record<string, any>): Promise<ApiResponse> {
    return this.request("POST", "accounts", data);
  }

  async updateAccount(
    id: number,
    data: Record<string, any>
  ): Promise<ApiResponse> {
    return this.request("PUT", `accounts/${id}`, data);
  }

  // ── Cadences ──
  async listCadences(params?: Record<string, string>): Promise<ApiResponse> {
    return this.request("GET", "cadences", undefined, params);
  }

  async getCadence(id: number): Promise<ApiResponse> {
    return this.request("GET", `cadences/${id}`);
  }

  // ── Cadence Memberships ──
  async addPersonToCadence(
    personId: number,
    cadenceId: number
  ): Promise<ApiResponse> {
    return this.request("POST", "cadence_memberships", {
      person_id: personId,
      cadence_id: cadenceId,
    });
  }

  async listCadenceMemberships(
    params?: Record<string, string>
  ): Promise<ApiResponse> {
    return this.request("GET", "cadence_memberships", undefined, params);
  }

  // ── Activities ──
  async listCalls(params?: Record<string, string>): Promise<ApiResponse> {
    return this.request("GET", "activities/calls", undefined, params);
  }

  async listEmails(params?: Record<string, string>): Promise<ApiResponse> {
    return this.request("GET", "activities/emails", undefined, params);
  }

  // ── Users ──
  async listUsers(params?: Record<string, string>): Promise<ApiResponse> {
    return this.request("GET", "users", undefined, params);
  }

  async getUser(id: number): Promise<ApiResponse> {
    return this.request("GET", `users/${id}`);
  }
}
